### ISSUE TITLE
[BUG/ISSUE] MODIS LAI have different values depending on number of cores used

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#1103]
### What institution are you from?
GCST

### Description of the problem
I have been tracing down a possible parallelization error.  It seems that when the HEMCO SoilNox extension is on we get slightly different values in in NO emissions from soil & fertilizer when running GEOS-Chem Classic 13.4.0 (dev branch) with different numbers of cores.  I chose Ref=5 cores and Dev=8 cores, which should cause a different ordering of the grid boxes in parallel loops.

```console
.. Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable               Ref=5 cores              Dev=8 cores              Dev - Ref
EmisNO_Fert          : 9.2054797e-10          | 9.198547e-10           | -6.93291615522923e-13 
EmisNO_Soil          : 3.152729e-09           | 3.1517353e-09          | -9.936293324588896e-13 
InvSOILNOX_NO        : 3.152729e-09           | 3.1517353e-09          | -9.936293324588896e-13 
EmisNO_Total         : 2.187116e-08           | 2.1870168e-08          | -9.93621200942596e-13 
```
This seems to be caused by the leaf area indices having different values when different core counts are used, at least for some land types.  I used the same executable in 2 different run directories, and ran Ref with 5 cores and Dev with 8.

Dev, on 8 cores
```console
State_Met%XLAI_NATIVE(13,38,41) = LAI for land type 41 = 1.3136272318661213E-002

State_Met%MODISLAI(13,38) = LAI summed over all land types = 3.4150421358644962  
```

Ref, on 5 cores
```console
State_Met%XLAI_NATIVE(13,38,41) = LAI for land type 41 = 1.0795551352202892E-002

State_Met%MODISLAI(13,38) = LAI summed over all land types = 3.4127014148980379     
```

Ideally, we should get identical results independent of core counts.  This could point to an issue with HEMCO regridding in GCClassic, at least for the leaf area indices.  I will keep investigating.

Tagging @lizziel @msulprizio @jimmielin

**Comment by yantosca** [comment]
Given that the code freeze date for 13.4.0 is upon us, we may need to fix this in a 13.4.Z version.

**Comment by yantosca** [comment]
@lizziel suggested to look at the `State_Met%LandTypeFrac`, which comes from the Olson map.  This can point to a problem in Olson map regridding.

From a new run I printed out the following:
```console
 >>> 5 cores: T, LANDTYPEFRAC(13,38,T)          41   3.2876322511583567E-003
 >>> 8 cores: T, LANDTYPEFRAC(13,38,T)          41   3.2876322511583567E-003

 >>> 5 cores: T, XLAI_NATIVE(13,38,T)           41   1.0795551352202892E-002
 >>> 8 cores: T, XLAI_NATIVE(13,38,T)           41   1.3136272318661213E-002
```
So for grid box (13,38) and landtype 41, the `State_Met%XLAI_Native` differs but the `State_Met%LandTypeFrac` is the same.  This would suggest that the Olson map regridding is OK but there is a problem in the LAI regridding and/or interpolation.

**Comment by yantosca** [comment]
A few issues affecting numerical precision in both GEOS-Chem and HEMCO have been fixed with commits c2d8a3da2fb043ae689b392c128101e498e8a18e and https://github.com/geoschem/HEMCO/commit/8afbe217ea2c112eb2a2dc1743852214b5535c94.  These fixes now make sure that XLAI_NATIVE coming out of HEMCO have the same values regardless of how many cores are used.

However there may be outstanding additional parallelization or numerical precision issues, and we are looking into this.  In particular, when MEGAN emissions are turned on, we get the following diffs (i.e. differences of totals) in HEMCO diagnostic output:

```console
Using configuration file hco.yml
... Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable               Ref=5 cores              Dev=8 cores              Dev - Ref
EmisACET_Biogenic    : 7.035722e-09           | 7.037491e-09           | 1.7689317892422007e-12 
EmisALD2_Biogenic    : 2.0391133e-09          | 2.0385378e-09          | -5.755726299218333e-13 
EmisC2H4_Biogenic    : 2.7339144e-09          | 2.7371352e-09          | 3.2210930971110496e-12 
EmisEOH_Biogenic     : 2.0391133e-09          | 2.0385378e-09          | -5.755726299218333e-13 
EmisHNO3_Ship        : 1.1806284e-09          | 1.1806284e-09          | 4.1152546494423447e-17 
EmisISOP_Biogenic    : 2.3401991e-08          | 2.3452106e-08          | 5.0114898342279446e-11 
EmisLIMO_Biogenic    : 1.6158672e-09          | 1.6166595e-09          | 7.922788944000891e-13 
EmisMOH_Biogenic     : 1.2897113e-08          | 1.2910582e-08          | 1.3470478205102054e-11 
EmisMTPA_Biogenic    : 1.23636354e-08         | 1.2368304e-08          | 4.6686977200893764e-12 
EmisMTPO_Biogenic    : 5.2480424e-09          | 5.250144e-09           | 2.1016758212227815e-12 
EmisNO2_Ship         : 1.5684917e-09          | 1.5684917e-09          | -1.424314686693423e-17 
EmisNO_Soil          : 2.947051e-09           | 2.9460285e-09          | -1.02221432273647e-12 
EmisPRPE_Biogenic    : 3.869745e-09           | 3.870967e-09           | 1.2224598503612905e-12 
EmisSOAP_Biogenic    : 1.3279525e-09          | 1.3290692e-09          | 1.1165554158337754e-12 
EmisSOAS_Biogenic    : 1.3279525e-09          | 1.3290692e-09          | 1.1165554158337754e-12 
InvIODINE_HOI        : 2.8343977e-10          | 2.8343977e-10          | 3.167056189783829e-18 
InvIODINE_I2         : 1.7113151e-11          | 1.7113151e-11          | 8.586465630545448e-20 
InvMEGAN_AAXX        : 3.4461994e-10          | 3.4453074e-10          | -8.922472949281918e-14 
InvMEGAN_ACET        : 7.035722e-09           | 7.037491e-09           | 1.7689317892422007e-12 
InvMEGAN_ACET_DIRECT : 6.864184e-09           | 6.8659647e-09          | 1.781180671706073e-12 
InvMEGAN_ACET_MBOX   : 1.6291553e-10          | 1.6291238e-10          | -3.1438957679070546e-15 
InvMEGAN_ALD2        : 2.0391133e-09          | 2.0385378e-09          | -5.755726299218333e-13 
InvMEGAN_APIN        : 7.209814e-09           | 7.2120865e-09          | 2.2724944755830867e-12 
InvMEGAN_BCAR        : 7.182023e-10           | 7.184082e-10           | 2.0600945808187804e-13 
InvMEGAN_BPIN        : 3.0099523e-09          | 3.011248e-09           | 1.2956800346172748e-12 
InvMEGAN_CARE        : 1.4667807e-09          | 1.467495e-09           | 7.142701686189967e-13 
InvMEGAN_EOH         : 2.0391133e-09          | 2.0385378e-09          | -5.755726299218333e-13 
InvMEGAN_FARN        : 5.8446864e-10          | 5.845989e-10           | 1.3027628292545251e-13 
InvMEGAN_FAXX        : 3.4461994e-10          | 3.4453074e-10          | -8.922472949281918e-14 
InvMEGAN_ISOP        : 2.3401991e-08          | 2.3452106e-08          | 5.0114898342279446e-11 
InvMEGAN_LIMO        : 1.6158672e-09          | 1.6166595e-09          | 7.922788944000891e-13 
InvMEGAN_MBOX        : 6.3250205e-10          | 6.326081e-10           | 1.0599548040494955e-13 
InvMEGAN_MOH         : 1.2897113e-08          | 1.2910582e-08          | 1.3470478205102054e-11 
InvMEGAN_MTPA        : 1.23636354e-08         | 1.2368304e-08          | 4.6686977200893764e-12 
InvMEGAN_MTPO        : 5.2480424e-09          | 5.250144e-09           | 2.1016758212227815e-12 
InvMEGAN_MYRC        : 9.368119e-10           | 9.371082e-10           | 2.963368753942236e-13 
InvMEGAN_OCIM        : 1.2506624e-09          | 1.2516645e-09          | 1.0022570880868775e-12 
InvMEGAN_OMON        : 3.063822e-09           | 3.0646483e-09          | 8.262062908825918e-13 
InvMEGAN_OSQT        : 1.3161396e-09          | 1.3164685e-09          | 3.287987829052319e-13 
InvMEGAN_PRPE        : 3.869745e-09           | 3.870967e-09           | 1.2224598503612905e-12 
InvMEGAN_SABI        : 6.827805e-10           | 6.831994e-10           | 4.1881122593394327e-13 
InvMEGAN_SOAP        : 1.3279525e-09          | 1.3290692e-09          | 1.1165554158337754e-12 
InvMEGAN_SOAS        : 1.3279525e-09          | 1.3290692e-09          | 1.1165554158337754e-12 
InvPARANOX_HNO3      : 1.1806284e-09          | 1.1806284e-09          | 4.1152546494423447e-17 
InvPARANOX_NO        : 2.5299218e-10          | 2.5299218e-10          | -1.076184310808028e-17 
InvPARANOX_NO2       : 1.5684917e-09          | 1.5684917e-09          | -1.424314686693423e-17 
InvSOILNOX_NO        : 2.947051e-09           | 2.9460285e-09          | -1.02221432273647e-12 
EmisACET_Total       : 1.5187526e-08          | 1.5189295e-08          | 1.7689395954978426e-12 
EmisALD2_Total       : 1.3417567e-08          | 1.3416992e-08          | -5.755779967225871e-13 
EmisC2H4_Total       : 6.096187e-09           | 6.0994076e-09          | 3.221101987568864e-12 
EmisEOH_Total        : 3.490591e-09           | 3.4900154e-09          | -5.755762077890025e-13 
EmisISOP_Total       : 2.340199e-08           | 2.3452106e-08          | 5.0114894872832494e-11 
EmisMOH_Total        : 1.7839762e-08          | 1.785323e-08           | 1.347050162386898e-11 
EmisMTPA_Total       : 1.4460515e-08          | 1.4465185e-08          | 4.668626162745992e-12 
EmisNO_Total         : 2.005815e-08           | 2.005713e-08           | -1.0221271528818021e-12 
EmisPRPE_Total       : 6.597229e-09           | 6.598451e-09           | 1.222458332478249e-12 
EmisSOAP_Total       : 8.684428e-09           | 8.685546e-09           | 1.1165427306683573e-12
```
These differences disappear when MEGAN biogenic emissions and chemistry are switched off.  We are still investigating.

**Comment by yantosca** [comment]
Found and fixed some unrelated parallelization and precision issues in `HEMCO/src/Extensions/hcox_soilnox_mod.F90`.  Long story short:

  1. Several numerical constants needed the kind parameter to force the proper precision (e.g. instead of 0.0 use 0.0_sp or 0.0_hp).
  2. We needed to zero the PRIVATE variables in this parallel DO loop: 

https://github.com/geoschem/HEMCO/blob/e8dfc811e161cc8b694e37985e58c27c6c97c5ec/src/Extensions/hcox_soilnox_mod.F90#L481-L526  

The fix now looks like this:
```F90
    ! Loop over each land grid-box, removed loop over landpoints
!$OMP PARALLEL DO                                                            &
!$OMP DEFAULT( SHARED )                                                      &
!$OMP PRIVATE( I, J, Dep_Fert, SoilFrt, FertDiag, IjFlux                    )&
!$OMP COLLAPSE( 2                                                           )
    DO J = 1, HcoState%NY
    DO I = 1, HcoState%NX

       ! Zero private variables for safety's sake
       Dep_Fert = 0.0_hp
       SoilFrt  = 0.0_hp
       FertDiag = 0.0_hp
       IJflux   = 0.0_hp
      ... etc...
```
I also took the opportunity to increase the number of grid boxes that will be parallelized over by using `!$OMP COLLAPSE(2)`.

This now gives identical results on different numbers of cores.  Further validation is ongoing.

**Comment by yantosca** [comment]
Still tracking what seems to be another issue in the computation of the leaf area indices via HEMCO:

Debug printout in `GeosCore/modis_lai_mod.F90`, at 4 x 5 grid box (22,10):

with 1 core:
```console
 MODISLAI:    9.5144298160448670E-002
 XLAI_NATIVE:            1   5.7757019996643066E-002
 XLAI_NATIVE:            3   1.0345440357923508E-002
 XLAI_NATIVE:            9   2.9938633088022470E-004
 XLAI_NATIVE:           13   4.6936993021517992E-004
 XLAI_NATIVE:           43   1.3133012689650059E-003
 XLAI_NATIVE:           55   2.4959780275821686E-002
```
with 5 cores:
```console
 MODISLAI:    9.4762627908494323E-002
 XLAI_NATIVE:            1   5.7757019996643066E-002
 XLAI_NATIVE:            3   1.0345440357923508E-002
 XLAI_NATIVE:            9   2.9938633088022470E-004
 XLAI_NATIVE:           13   4.6936993021517992E-004
 XLAI_NATIVE:           43   9.3163101701065898E-004
 XLAI_NATIVE:           55   2.4959780275821686E-002
```
with 8 cores:
```console
 MODISLAI:    9.5144298160448670E-002
 XLAI_NATIVE:            1   5.7757019996643066E-002
 XLAI_NATIVE:            3   1.0345440357923508E-002
 XLAI_NATIVE:            9   2.9938633088022470E-004
 XLAI_NATIVE:           13   4.6936993021517992E-004
 XLAI_NATIVE:           43   1.3133012689650059E-003
 XLAI_NATIVE:           55   2.4959780275821686E-002
```
MODISLAI is the sum of XLAI_NATIVE over all land types.  As you can see the land type #43 has different values.  I believe that this is happening in HEMCO routine Get_Current_Emisisons.  Investigation is continuing.

**Comment by yantosca** [comment]
We now know the cause of this issue.  It is not a parallelization problem per se, but HEMCO was reading in the incorrect time-slice for MODIS LAI as explained in https://github.com/geoschem/HEMCO/issues/132.

**Comment by yantosca** [comment]
We can close this issue as the solution is provided in https://github.com/geoschem/HEMCO/issues/132.

**Comment by yantosca** [comment]
Remaining parallelization issues have been solved in PR #1190.
